### ISSUE TITLE
Mandatory vs. optional parentheses for function calls

### ISSUE THREAD

**Issue by faultyserver** [myst-lang/myst#8]
I'm a bit torn over whether or not to require parentheses to designate function calls.

## Background and similar languages

- Ruby has always treated parentheses as always-optional.
- Crystal only requires parentheses for definitions of functions that take parameters
- Elixir [changed in 1.4](https://github.com/elixir-lang/elixir/issues/3268) to require parentheses for zero-argument calls, but not for multi-argument calls.

Myst is not really a message-passing language. As such, there's a bit of a distinction between a member access and a function call. Right now, a member function call looks like two separate nodes: an `AccessExpression` for the `object.member` part, and a `FunctionCall` for the `member()` part.

This complicates things a bit. `FunctionCall` has to know about `object` as the receiver in the `object.member()` structure. As a separate node, it can't really know that without some funky stack stuff and guessing from the interpreter/hints from the function being called.

I dislike the inconsistency that Elixir _had_ where parentheses were optional for zero-argument calls. And still dislike that they are not required for calls with arguments. It's better, but still confusing to read if you are unfamiliar with the context.

I currently see two options, neither of which are really ideal to me.

## Option 1

Always require parentheses for function calls.

This is the current behavior. In this case, those two node instances can be merged to a single `MemberFunctionCall` that has the entire context of `receiver.member()`. This also still allows for chaining like `something.split().sum()`.

In cases like that, however, I specifically appreciate the ability to omit parentheses (`something.split.sum` looks a lot better in my eyes).

## Option 2

Make parentheses fully optional.

Doing so would cause member accesses that result in functors to _always_ be evaluated immediately. This means obtaining a reference to a member function would require additional syntax (currently `ref = object.member_fun` will assign `ref` as a bound functor that can later be called as `ref(args...)`. This doesn't work if the functor would be immediately evaluated).

Something I really like from C/C++ is the ability to plainly obtain function references by writing their name with no call notation. This does, however, make `something.split.sum` immediately work.

------

I would really like a 3rd option that compromises between the two. I _might_ be okay with more syntax to obtain references. I am _not_ okay with requiring bare parentheses everywhere.

**Comment by faultyserver** [comment]
A quick idea I had this morning was using a different access notation to indicate functor reference instead of adding a prefix notation. This would be similar to `.` vs `->` in C/C++. Example:

```crystal
reference = object->func
call_result = object.func
```

I don't particularly like `->`, mostly because it is two characters wide (there is almost certainly a single character that would look better) and is unintuitive for anyone familiar with C/C++ or PHP.

I _do_, however, like that it avoids a prefix notation for the reference. Particularly when doing deep references, it can be confusing to have to look through lots of characters to know that it is a reference and not a call.

Alternatively, a suffix notation may also work well:

```crystal
ref = object.func&
call = object.func
```

Again, I'm not entirely sold on the `&` character, but it works as an example and is actually consistent with how blocks are passed normally:

```crystal
def func(thing, &block)
end

func(thing, &other_func)
```

**Comment by faultyserver** [comment]
I could see using either `#` or `::` for this. Both are standard fare when referencing functions _outside_ of code (e.g., [Ruby's documentation](http://ruby-doc.org/core-2.4.1/Array.html)).

Since namespacing is currently only done through `.`, (unlike Ruby where something like `Token::Type` would namespace, Myst would use `Token.Type`), the `::` operator is currently free. It's use could be something like:

```ruby
ref = object.member::func
```

`#` would work similarly. The only reason _not_ to do this is that it overloads it's meaning from a comment and would likely trip up most editors:

```ruby
ref = object.member#func
```

I like the style and differentiation of `#` more, but I think `::` will be more usable.

**Comment by faultyserver** [comment]
I'm liking the `::` notation more and more. It's a slight overload of the operator from the C family, but the semantics of that aren't too ingrained to make it infeasible. I'm going to ignore the `#` notation from now on. If comments used a different syntax, it might have been considered.

With that, for top-level or in-scope references, this infix notation would work well by appearing as a prefix notation, but restricts the ability to look at higher scopes.

```ruby
def func
  IO.puts(1)
end

module Scope
  def func
    IO.puts(2)
  end

  # This would refer to the func defined inside of `Scope`, counterintuitive to
  # how almost all other scoped languages work.
  ref_to_scoped_func = ::func
end

# This also refers to `Scoped#func`.
ref_to_scoped_func = Scoped::func
# This refers to the top-level `func`.
ref_to_func = ::func
```

I'm liking this syntax overall, but it's semantics need to be thought out more and cases like this will need to be handled (How can expressions inside of `Scoped` access the outside `func`?)

**Comment by faultyserver** [comment]
Back to the original purpose of this issue:

I think the solution that #14 will bring is to make parentheses optional for zero-argument calls, but _always required_ for calls with one or more arguments.

This might end up being a little annoying for DSLs (thinking of things like `belongs_to` and `has_many` from Rails), but consistency and clarity are more important to me here.

Specifically, this avoids things like `call some thing, with, args`, which in Ruby works out to `call(some(thing, with, args))`, but at a glance, might be interpreted as `call(some, thing, with, args)`.

Block arguments don't count in the arguments mentioned above. That is, something like `each{ do_thing }` is valid. As another bonus, the required parentheses for calls means that this _cannot be ambiguous_ with a call with a Map argument. For example, in Ruby, `puts {}` _looks like_ a puts with an empty Hash, but is _interpreted_ as a puts with an empty _block_.

```ruby
irb(main):004:0> puts {}

=> nil
irb(main):005:0> puts({})
{}
=> nil
```

This has been implemented as of 30c0683.

**Comment by faultyserver** [comment]
This was handled by #14 as mentioned above.
